First draft of packaging tooling

rdbisme opened the pull request:
This installs the python package and the extension doing: `pip install .`

## Problems: 
```
ImportError: /home/rdb/.local/mambaforge/envs/pgeon-dev/lib/python3.10/site-packages/pgeon/_pgeon.so: undefined symbol: _ZNK5arrow8DataType18ComputeFingerprintB5cxx11Ev
```
I think that the conda shipped `arrow-cpp` is not binary compatible with `pyarrow` stuff?

## Improvement Ideas: 
- Edit the cmake stuff to link against `pyarrow` installed libs instead of `arrow-cpp` such that installing arrow in the system is not required anymore when installing via Python (keep existing behaviour for C++ only compilation)
- Add `pgeonConfig.cmake` and targets and related stuff to make it a well behaving CMake package tool.

0x0L commented on src/python/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+add_library(cy_pgeon MODULE ${cy_pgeon})
+set_target_properties(cy_pgeon PROPERTIES PREFIX "" OUTPUT_NAME "_pgeon")
+
+target_link_libraries(cy_pgeon PRIVATE pgeon arrow_shared ${PYTHON_LIBRARY})
```
I think we are missing `${ARROW_PYTHON_SHARED_LIB}`
See `https://github.com/apache/arrow/blob/master/cpp/cmake_modules/FindArrowPython.cmake`

rdbisme replied:
I hardcoded the `ArrowPython` path in the new revision I just pushed, and the `_pgeon.so` is correctly linked against `arrowpython`: 
```
~/git/pgeon/dist/pgeon]$ ldd _pgeon.so 
        linux-vdso.so.1 (0x00007ffdcb7fb000)
        libarrow.so.700 => /home/rdibattista/mambaforge/envs/pgeon-dev/lib/libarrow.so.700 (0x000075470467c000)
        libarrow_python.so.700 => /home/rdibattista/mambaforge/envs/pgeon-dev/lib/libarrow_python.so.700 (0x0000754705f95000)
```

but I still get the problem with the missing symbols: 
```
~/git/pgeon/dist/pgeon]$ python -c "from pgeon import *"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/rdibattista/mambaforge/envs/pgeon-dev/lib/python3.10/site-packages/pgeon/__init__.py", line 3, in <module>
    from ._pgeon import copy_query
ImportError: /home/rdibattista/mambaforge/envs/pgeon-dev/lib/python3.10/site-packages/pgeon/_pgeon.so: undefined symbol: _ZNK5arrow8DataType18ComputeFingerprintB5cxx11Ev
```

Some investigation: 
```
find /home/rdibattista/mambaforge/envs/pgeon-dev/lib -type f -name "*.so" -exec sh -c  "objdump -T {}  | grep  '_ZNK5arrow8DataType18ComputeFingerprint'" \;
0000000000000000      DF *UND*  0000000000000000              _ZNK5arrow8DataType18ComputeFingerprintEv
0000000000000000      DF *UND*  0000000000000000              _ZNK5arrow8DataType18ComputeFingerprintB5cxx11E
```

That `arrow8` in the string makes me think it's somehow getting `arrow8` and not arrow7`

rdbisme replied:
I think the solution is to link against the python wheel stuff only when building on Python. I'll write a custom `findArrow.cmake` module.

rdbisme commented:
Switched to `pybind11` and this should work as expected (did not test runtime, but I don't have symbols problems anymore).

```
mamba env create -f environment.yml
mamba activate pgeon-dev
pip install .
```

I just exported the C++ `CopyQuery` to `copy_query`. Dunno if there's a bit of argument recasting needed to make it work as expected from Python.

rdbisme commented:
Ok, this compiles and installs the package without interaction in the devcontainer. 

Now I tried to run the `t.py` after populating the db and I'm getting a segmentation fault.

rdbisme commented:
Added a separate `build-environment.yml` for future usage to build the wheel in CI. I thought about building also the shared lib together with the python wrapper, and I think we should not do it. If people are interested to work on the C++ lib, they can build it manually using `cmake`. They can always trigger the build of the wrapper doing `pip install .` 

Also, adding the two different targets when `SKBUILD` makes IMHO the `CmakeLists.txt` unnecessary clumsy. 

Also updated the documentation. 

I'll work on refactoring tests in standard `pytest` suite in another PR with CI.
